[processors/routingprocessor] Extract signals processing logic from router.

kovrus opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

Currently, `router.go` performs all the logic related to registering exporters and signals processing. Now, it also combines generic and concrete types. After prototyping the solution [#13158](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13158) it was obvious that it is hard to extend the processor with it's current state. This change does the following:
- Splits router logic into three signal processors and generic routing logic. Processors are responsible for consuming signals and router for registering default and requested exporters and further used in processors. 
- Removes [intermediate step](https://github.com/open-telemetry/opentelemetry-collector-contrib/compare/main...kovrus:issue-13158-split-signals?expand=1#diff-58ff3d2882be24fdb4db5369e8929f033a346e5b5caca85e501d6585d5ef6356L126-L138) after grouping signal entries by routing value and make it possible to consume groups without allocating additional slice to hold them. 

 
**Link to tracking Issue:** <Issue number if applicable>
[13158](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13158)

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

kovrus committed:
[chore] Create dedicated signal routing processors.

jpkrohling commented on processor/routingprocessor/factory_test.go:
```diff
@@ -234,3 +234,20 @@ type mockProcessor struct{}
 	return ptrace.NewTraces(), nil
 }
+
+type mockHost struct {
```
One thing I always wished was that our noop component test helpers would be more like mocks. Perhaps that would be a good contribution to the core repo?

https://github.com/open-telemetry/opentelemetry-collector/blob/e092fc728b724f862d37294dff03a602996f1dd2/component/componenttest/nop_host.go#L26-L28

kovrus replied:
I can open a ticket for that.

jpkrohling commented on processor/routingprocessor/logs.go:
```diff
@@ -0,0 +1,146 @@
+	router    router[component.LogsExporter]
+}
+
+// newProcessor creates new processor
```
```suggestion
// newLogProcessor creates new processor
```

jpkrohling replied:
Or just get rid of this comment, it's not that useful anyway.

jpkrohling reviewed (approved):
I like the new split. There's some code duplication (like before), but I think it's justifiable.

kovrus commented:
Squashed commits and also dropped https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13622/files#diff-58ff3d2882be24fdb4db5369e8929f033a346e5b5caca85e501d6585d5ef6356L326

kovrus commented:
cc: @gouthamve

jpkrohling commented:
> Squashed commits and ...

Adding new commits to the PR is usually better, as it makes it easier to perform subsequent reviews. This way, we can see what's new between reviews. When we merge, we can squash the commits.

kovrus committed:
Fix linting errors.

kovrus commented:
@jpkrohling, I just fixed the linting errors, can you take another look?
